`DimeNet++` implementation

arunppsg opened the pull request:
In this PR, I implemented DimeNet++ by subclassing DimeNet. The `InteractionPPBlock` and `OutputPPBlock` was based on this [implementation](https://github.com/Open-Catalyst-Project/ocp/blob/main/ocpmodels/models/dimenet_plus_plus.py).

Also, I depreciated `OutputBlock`, thereby DimeNet model will also be using `OutputPPBlock`. `OutputPPBlock` allows up-projection and down-projection of embeddings, thereby removing information bottlenecks.

Fixes #4427

rusty1s reviewed (commented):
Thanks for the PR. Can we add a basic test for `DimeNetPlusPlus` as well? Also I am interested in whether it is possible to make use of pre-trained weights for this model as well (similar to what we do for `DimeNet`), see [here](https://github.com/gasteigerjo/dimenet/tree/master/pretrained/dimenet_pp).

arunppsg commented:
Sure, I will look into using pre-trained weights and also add tests.

Padarn commented on torch_geometric/nn/models/dimenet.py:
```diff
@@ -242,11 +334,23 @@ def reset_parameters(self):
+                         out_channels, num_layers, act=swish)
+        import warnings
+        warnings.warn(
+            "OutputBlock will be depreciated in PyG 2.1.0. Use OutputPPBlock instead",  # noqa
```
Any reason we don't use `OutputPPBlock` this will be depreciated?

arunppsg replied:
The DimeNet++ paper says that OutputPPBlock has less information bottlenecks as it allows for information up projection and down projection of embedings. The difference between them is minor and I thought that OutputPPBlock will also benefit DimeNet model users. Hence the change.

But I have to test whether OutputPPBlock will be compatible with DimeNet when using pre-trained weights. If it is not compatible, I will revert back.

rusty1s replied:
Thank you. On that note, it might makes sense to add a test for loading pre-trained weights in general. Can try to do this in a separate PR.

arunppsg replied:
@Padarn I have restored OutputBlock as OutputPPBlock was not compatible with DimeNet when using pretrained weights.

@rusty1s , I would like to add tests in a separate PR.

Padarn replied:
Thanks for the update @arunppsg. Once this is merged could you make an issue describing the `OutputPPBlock` problem? 🙇

arunppsg replied:
I don't think it's a problem - earlier DimeNet was using OutputBlock, I thought it can use OutputPPBlock since it is more efficient but realized that it will not be compatible, then reverted back. So, users who want to use OutputPPBlock can directly use DimeNet++

Padarn replied:
Oh I see now, sorry I got mixed up because I was reading it via email. Thanks.

arunppsg commented:
@rusty1s regarding tests for DimeNetPlusPlus, I have added a test for checking output size and another one test for checking whether the model is training well or not - test_overfit. The downside of test_overfit is that it is sometimes flaky but otherwise it is a good test for checking whether the model is able to train or not.

rusty1s commented:
Thank you @arunppsg! I am currently on vacation, will have a final look in the next week.

arunppsg commented:
The failures in testing are due to not having `sympy` in requirements. It is used by DimeNet++ and DimeNet for basis computation.

rusty1s commented:
@arunppsg Can we add the necessary dependencies to `full_install_requires`, see [here](https://github.com/pyg-team/pytorch_geometric/blob/master/setup.py#L22-L35)?

arunppsg commented:
@rusty1s added the necessary requirements.

rusty1s commented:
Thanks @arunppsg. I think I fixed the dependency issue during testing. I also added a `--use_dimenet_plus_plus` option to the QM9 script. However, the results of the pre-trained model are far off, e.g.:
```
Target: 00, MAE: 4.85522 ± 3.42503
Target: 01, MAE: 15.25113 ± 6.21239
Target: 02, MAE: 603.06146 ± 484.24606
Target: 03, MAE: 994.61578 ± 703.35645
Target: 05, MAE: 977.33118 ± 332.41879
Target: 06, MAE: 594.56433 ± 317.66690
Target: 07, MAE: 3095.07080 ± 1889.53760
```
Do you know what's the reason for this?

arunppsg commented:
> Thanks @arunppsg. I think I fixed the dependency issue during testing. I also added a `--use_dimenet_plus_plus` option to the QM9 script. However, the results of the pre-trained model are far off, e.g.:
> 
> ```
> Target: 00, MAE: 4.85522 ± 3.42503
> Target: 01, MAE: 15.25113 ± 6.21239
> Target: 02, MAE: 603.06146 ± 484.24606
> Target: 03, MAE: 994.61578 ± 703.35645
> Target: 05, MAE: 977.33118 ± 332.41879
> Target: 06, MAE: 594.56433 ± 317.66690
> Target: 07, MAE: 3095.07080 ± 1889.53760
> ```
> 
> Do you know what's the reason for this?

Not sure about it, the results are far off from the dimenet++ paper. Can we create an issue and track it seperately?

arunppsg commented on torch_geometric/nn/models/dimenet.py:
```diff
@@ -454,8 +454,8 @@ def reset_parameters(self):
 
-    @staticmethod
-    def from_qm9_pretrained(root: str, dataset: Dataset, target: int):
+    @classmethod
```
Just curious to know, how is using classmethod different from using staticmethod here?

rusty1s replied:
I think this makes more sense here since it can naturally handle inheritance (`cls.url` except for `DimeNet.url`).

rusty1s commented:
This sounds good to me. I will try to look into this as well.

arunppsg commented:
Thank you!
